### PR TITLE
Product hints

### LINKED ISSUE #68: Default ConfigFieldMapping should be (CamelCase, KebabCase)
The current default mapping between config keys and case class fields is camel case to camel case. However, in Typesafe config the _de facto_ standard is to use dashes between words (see examples in [Typesafe config](https://github.com/typesafehub/config#uses-of-substitutions), [akka](http://doc.akka.io/docs/akka/snapshot/general/configuration.html#Listing_of_the_Reference_Configuration), [spray](https://github.com/spray/spray/blob/master/spray-can/src/main/resources/reference.conf), [kamon](https://github.com/kamon-io/Kamon/blob/master/kamon-core/src/main/resources/reference.conf) and many more). As such, it seems more natural to me for the default to be `ConfigFieldMapping[T](CamelCase, KebabCase)`.

What is your opinion on this? Is there a reason for the current default? If you agree with that change I can do a quick pull request.

### PR DESCRIPTION
**Pull request by ruippeixotog**
This is an implementation of a `ProductHint`, as discussed in #76. It supports:

- Defining the mapping between field names and config keys (with `ConfigFieldMapping`);
- Controlling whether default arguments should be applied or not;
- Controlling whether unknown keys are allowed or not.

For easier reviewing, I added all those features in different commits.

Some random details:

- As much as I'd like it, there is a major problem with enabling "strict mode" as default: configs that are loaded using `ConfigFactory.load` and are read from the root will always have unknown keys, as Typesafe config mixes in environment variables and system properties;
- I could have created an implicit conversion from `ConfigFieldMapping` to `ProductHint` in order not to break existing code using implicit `ConfigFieldMappings`. However, that will lead to confusion in the long run if we want to use mappings for other things, like coproduct fields. I personally prefer to have users deal with breaking changes if the alternative is to introduce hacks and somewhat magical behaviors that will hinder future improvements;
- I'd like to take this opportunity to close #68, if we're all ok with it. If we're already changing the API for `ConfigFieldMapping` we could make users also do the transition to a new naming convention, dealing with breaking code only once;
- I did not touch README and CHANGELOG in this pull request to avoid being repeatedly updating it as changes are requested. I prefer to do a separate pull request updating all documentation once this is closed.

### CODE REVIEW

**Review by melrief — APPROVED** [review]
@ruippeixotog I had a look at this and I have no comments. Ready to merge but I would prefer if either @leifwickland or @jcazevedo can have a look at this and approve it as well. Thanks for doing this.

File: core/src/main/scala/pureconfig/ProductHint.scala
Comment (jcazevedo) [inline core/src/main/scala/pureconfig/ProductHint.scala]: Do you think it still makes sense for `ConfigFieldMapping` to exist as an entity? I think we could simplify the interface if `ProductHint` only had a `def configFieldName(typeFieldName: String): String` method. We should still provide helpers to create `ProductHint`s from two `NamingConvention`s, as is currently available for `ConfigFieldMapping`s.

**Comment by jcazevedo** [comment]
I'm also in favor of closing #68. If we are to change the default, now is a good time to do it.

**Comment by melrief** [comment]
I agree on fixing #68, please @ruippeixotog proceed

File: core/src/main/scala/pureconfig/ProductHint.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/ProductHint.scala]: I agree that without its the implicit usage, `ConfigFieldMapping` is not much more than a `String => String`. Having a way to create such a function object - at any time, and not only via a `ProductHint` - still seems useful, though. First, you can use an independent `ConfigFieldMapping` to manually convert fields when you are implementing a custom `ConfigConvert`. Second, it is slightly more elegant to pass a function to the `ProductHint` constructor than passing two `NamingConvention`s without extra context.

How about we leave the `ConfigFieldMapping` object and make its `apply` methods simply return `String => String`? We could then accept any `String => String` function in hints and use the interface you proposed. The downside is less context - when a `ProductHint` receives a string function, how is it applied? Is it a function from config keys to fields or viceversa?

**Comment by ruippeixotog** [comment]
Ok! Changed the default naming convention in 01ce8e7.

**Comment by melrief** [comment]
Thanks @ruippeixotog. I'm waiting for one more review and then merge it, good job.

File: core/src/test/scala/pureconfig/PureconfSuite.scala
Comment (leifwickland) [inline core/src/test/scala/pureconfig/PureconfSuite.scala]: Suggestion: Update this test's name to indicate that it is testing CamelCase instead of word delimiters.

**Review by leifwickland — APPROVED** [review]
Great work, @ruippeixotog!
Please consider sending a PR to update the README to reflect the new way of customizing field mappings and the ability to influence whether default values are accepted.

File: core/src/main/scala/pureconfig/ProductHint.scala
Comment (jcazevedo) [reply core/src/main/scala/pureconfig/ProductHint.scala]: I think we can leave the `ConfigFieldMapping` and the `apply` method of `ProductHint` as is, and delegate the call in the default `ProductHintImpl` to the `ConfigFieldMapping`, no? Something along the lines of (if the method in `ProductHint` would be called `configFieldName`):

```scala
private[pureconfig] case class ProductHintImpl[T](
    fieldMapping: ConfigFieldMapping,
    useDefaultArgs: Boolean,
    allowUnknownKeys: Boolean) extends ProductHint[T] {
  def configFieldName(typeFieldName: String) = fieldMapping(typeFieldName)
}
```

**Comment by melrief** [comment]
I and @leifwickland approved this PR. @jcazevedo what about you?

File: core/src/main/scala/pureconfig/ProductHint.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/ProductHint.scala]: Ok, I agree with that. @melrief and @leifwickland, do you agree too?

**Comment by melrief** [comment]
@ruippeixotog merging #94 changed the style of `ConfigConvert` in master and that's why there is a conflict. It shouldn't be too difficult to fix it. I'll wait for the last review and merge this.

File: core/src/test/scala/pureconfig/PureconfSuite.scala
Comment (ruippeixotog) [reply core/src/test/scala/pureconfig/PureconfSuite.scala]: Well, the intention of this test is not to test specifically the camel case conversion - that is done extensively in `NamingConventionTest.scala`. This one is for testing that adding an appropriate hint allows different field mappings to be applied. However, I can replace "word delimiters" with "different naming conventions" and test more than one mapping. What do you think?

**Comment by ruippeixotog** [comment]
@leifwickland, I intend to update the README and the CHANGELOG once this is merged :)

File: core/src/test/scala/pureconfig/PureconfSuite.scala
Comment (leifwickland) [reply core/src/test/scala/pureconfig/PureconfSuite.scala]: That seems great!

File: core/src/main/scala/pureconfig/ProductHint.scala
Comment (leifwickland) [reply core/src/main/scala/pureconfig/ProductHint.scala]: _shrug_ Either way.

**Comment by jcazevedo** [comment]
> I and @leifwickland approved this PR. @jcazevedo what about you?

I'd rather not have the `ConfigFieldMapping` entity in the `ProductHint` trait, as suggested, but I'm OK with leaving this as it is right now if you guys prefer it.

File: core/src/test/scala/pureconfig/PureconfSuite.scala
Comment (ruippeixotog) [reply core/src/test/scala/pureconfig/PureconfSuite.scala]: Added in ae169e040ad5935fca410c27c9ec2f6496854230.

File: core/src/main/scala/pureconfig/ProductHint.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/ProductHint.scala]: Changed it in c84305a3253abf7c04432038c6d8090923050d06. @melrief, let me know if you prefer the other way.

**Review by jcazevedo — APPROVED** [review]
LGTM

**Comment by melrief** [comment]
Nice one, thanks!

**Pull request merged**
